packages/netlify/src/index.ts

```diff
@@ -1,2 +1,220 @@
+	netlify: {
+		context: Context;
+	};
+}
+
+const isStaticRedirect = (route: RouteData) =>
+	route.type === 'redirect' && (route.redirect || route.redirectRoute);
+
+const clearDirectory = (dir: URL) => rm(dir, { recursive: true }).catch(() => {});
+
+export interface NetlifyIntegrationConfig {
+	/**
+	 * If enabled, On-Demand-Rendered pages are cached for up to a year.
+	 * This is useful for pages that are not updated often, like a blog post,
+	 * but that you have too many of to pre-render at build time.
+	 *
+	 * You can override this behavior on a per-page basis
+	 * by setting the `Cache-Control`, `CDN-Cache-Control` or `Netlify-CDN-Cache-Control` header
+	 * from within the Page:
+	 *
+	 * ```astro
+	 * // src/pages/cached-clock.astro
+	 * Astro.response.headers.set('CDN-Cache-Control', "public, max-age=45, must-revalidate");
+	 * ---
+	 * <p>{Date.now()}</p>
+	 * ```
+	 */
+	cacheOnDemandPages?: boolean;
+}
+
+export default function netlifyIntegration(
+	integrationConfig?: NetlifyIntegrationConfig
+): AstroIntegration {
+	let _config: AstroConfig;
+	let outDir: URL;
+	let rootDir: URL;
+	let astroMiddlewareEntryPoint: URL | undefined = undefined;
+
+	const ssrOutputDir = () => new URL('./.netlify/functions-internal/ssr/', rootDir);
+	const middlewareOutputDir = () => new URL('.netlify/edge-functions/middleware/', rootDir);
+
+	const cleanFunctions = async () =>
+		await Promise.all([clearDirectory(middlewareOutputDir()), clearDirectory(ssrOutputDir())]);
+
+	async function writeRedirects(routes: RouteData[], dir: URL) {
+		const fallback = _config.output === 'static' ? '/.netlify/static' : '/.netlify/functions/ssr';
+		const redirects = createRedirectsFromAstroRoutes({
+			config: _config,
+			dir,
+			routeToDynamicTargetMap: new Map(
+				routes
+					.filter(isStaticRedirect) // all other routes are handled by SSR
+					.map((route) => {
+						// this is needed to support redirects to dynamic routes
+						// on static. not sure why this is needed, but it works.
+						route.distURL ??= route.redirectRoute?.distURL;
+
+						return [route, fallback];
+					})
+			),
+		});
+
+		if (!redirects.empty()) {
+			await appendFile(new URL('_redirects', outDir), '\n' + redirects.print() + '\n');
+		}
+	}
+
+	async function writeSSRFunction() {
+		await writeFile(
+			new URL('./ssr.mjs', ssrOutputDir()),
+			`
+				import createSSRHandler from './entry.mjs';
+				export default createSSRHandler(${JSON.stringify({
+					cacheOnDemandPages: Boolean(integrationConfig?.cacheOnDemandPages),
+				})});
+				export const config = { name: "Astro SSR", generator: "@astrojs/netlify@${packageVersion}", path: "/*", preferStatic: true };
+			`
+		);
+	}
+
+	async function writeMiddleware(entrypoint: URL) {
+		await mkdir(middlewareOutputDir(), { recursive: true });
+		await writeFile(
+			new URL('./entry.mjs', middlewareOutputDir()),
+			`
+						import { onRequest } from "${fileURLToPath(entrypoint).replaceAll('\\', '/')}";
```

Comment:
When dealing with Windows, we found that the best compatible way is to use `.toString()` instead of `replaceAll`. I think should be able to pick up the `file://` protocol. Also, does `entrypoint` contain the `C:\\` in this case?

Reply:
In my local testing, `.toString()` didn't work - it trips over the `file://` protocol. The current code seems to pass tests on windows - could you outline a scenario where this breaks?

Reply:
Oh, never mind. I just noticed now that you used `fileURLToPath`. In some other instance (vercel adapter and the previous of netlify adapter maybe) I used `URL#toString()`. I'd say it should be fine :)